modules/transactions/src/main/java/org/apache/ignite/internal/tx/storage/state/rocksdb/TxStateRocksDbStorage.java

```diff
@@ -479,40 +515,52 @@ private UUID keyToTxId(byte[] bytes) {
 
     @Override
     public void close() {
-        close0();
+        tryToCloseStorageAndResources();
     }
 
     @Override
     public CompletableFuture<Void> startRebalance() {
-        if (!STATE.compareAndSet(this, StorageState.RUNNABLE, StorageState.REBALANCE)) {
-            throwExceptionIfStorageClosedOrRebalance();
+        CompletableFuture<Void> rebalanceFuture = new CompletableFuture<>();
+
+        if (!REBALANCE_FUTURE.compareAndSet(this, null, rebalanceFuture)) {
+            throw createStorageInProgressOfRebalanceException();
         }
 
-        busyLock.block();
+        try {
+            if (!STATE.compareAndSet(this, StorageState.RUNNABLE, StorageState.REBALANCE)) {
+                throwExceptionIfStorageClosedOrRebalance();
+            }
 
-        try (WriteBatch writeBatch = new WriteBatch()) {
-            writeBatch.deleteRange(partitionStartPrefix(), partitionEndPrefix());
-            writeBatch.put(lastAppliedIndexAndTermKey, indexAndTermToBytes(REBALANCE_IN_PROGRESS, REBALANCE_IN_PROGRESS));
+            busyLock.block();
 
-            db.write(writeOptions, writeBatch);
+            try (WriteBatch writeBatch = new WriteBatch()) {
+                writeBatch.deleteRange(partitionStartPrefix(), partitionEndPrefix());
+                writeBatch.put(lastAppliedIndexAndTermKey, indexAndTermToBytes(REBALANCE_IN_PROGRESS, REBALANCE_IN_PROGRESS));
 
-            lastAppliedIndex = REBALANCE_IN_PROGRESS;
-            lastAppliedTerm = REBALANCE_IN_PROGRESS;
-            persistedIndex = REBALANCE_IN_PROGRESS;
+                db.write(writeOptions, writeBatch);
 
-            CompletableFuture<Void> rebalanceFuture = completedFuture(null);
+                lastAppliedIndex = REBALANCE_IN_PROGRESS;
+                lastAppliedTerm = REBALANCE_IN_PROGRESS;
+                persistedIndex = REBALANCE_IN_PROGRESS;
 
-            this.rebalanceFuture = rebalanceFuture;
+                rebalanceFuture.complete(null);
 
-            return rebalanceFuture;
-        } catch (Exception e) {
-            throw new IgniteInternalException(
-                    TX_STATE_STORAGE_REBALANCE_ERR,
-                    IgniteStringFormatter.format("Failed to clear storage for partition {} of table {}", partitionId, getTableName()),
-                    e
-            );
-        } finally {
-            busyLock.unblock();
+                return rebalanceFuture;
+            } catch (Exception e) {
+                throw new IgniteInternalException(
+                        TX_STATE_STORAGE_REBALANCE_ERR,
+                        IgniteStringFormatter.format("Failed to clear storage for partition {} of table {}", partitionId, getTableName()),
+                        e
+                );
+            } finally {
+                busyLock.unblock();
+            }
+        } catch (IgniteInternalException e) {
```

Comment:
Is this the exception from the line 550? Are there other sources for such type of exceptions? What do we do with the state?

Reply:
This is the handling of exceptions from lines 531 and 550. Yes, it is necessary to deal with standing, I will correct it a little.